[clang] Fix nullptr dereference when checking friend default comparison

BruceAko opened the pull request:
Fix a crash in `SemaDeclCXX.cpp` when defaulting a comparison operator that has been declared as a friend function alongside class friend declarations. When checking if a defaulted operator is a friend, the code would dereference the result of `FriendDecl::getFriendDecl()` without checking for nullptr, which occurs for class friend declarations.

The fix ensures we check that `getFriendDecl()` returns a non-null pointer before attempting to access its canonical declaration.

Fixes https://github.com/llvm/llvm-project/issues/132249

mizvekov commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -9039,8 +9039,8 @@ bool Sema::CheckExplicitlyDefaultedComparison(Scope *S, FunctionDecl *FD,
     if (llvm::none_of(RD->friends(), [&](const FriendDecl *F) {
-          return FD->getCanonicalDecl() ==
-                 F->getFriendDecl()->getCanonicalDecl();
+          NamedDecl *ND = F->getFriendDecl();
+          return ND && FD->getCanonicalDecl() == ND->getCanonicalDecl();
```
```suggestion
          NamedDecl *ND = F->getFriendDecl();
          return declaresSameEntity(F->getFriendDecl(), FD;
```

mizvekov reviewed (changes requested):
This should add test and release note.

BruceAko committed:
[clang] Fix nullptr dereference when checking friend default comparison

Fix comparison check for explicitly defaulted functions by using declaresSameEntity for friend declarations.

shafik reviewed (commented):
@BruceAko why did you close the PR?
